src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -3,122 +3,162 @@
         {
-            var result = new List<EndpointInstanceTimings>();
-            var intervals = GenerateIntervalIds(now, NumberOfHistoricalIntervals);
+            const int Size = NumberOfHistoricalIntervals*2;
+            ReaderWriterLockSlim rwl = new ReaderWriterLockSlim();
+            MeasurementInterval[] intervals = new MeasurementInterval[Size];
 
-            foreach (var instanceId in timings.Keys)
+            // ReSharper disable once SuggestBaseTypeForParameter
+            public void ReportTimeIntervals(DateTime now, TimingInterval[] intervalsToFill, out long totalDuration, out long totalMeasurements)
             {
-                ConcurrentDictionary<DateTime, MeasurementInterval> endpointInstanceData;
+                var epoch = GetEpoch(now.Ticks);
+                var howMany = intervalsToFill.Length;
 
-                if (timings.TryGetValue(instanceId, out endpointInstanceData))
-                {
-                    var totalDuration = 0L;
-                    var totalMeasurements = 0L;
-                    var timingIntervals = new TimingInterval[intervals.Length];
+                totalDuration = 0;
+                totalMeasurements = 0;
 
-                    for (var i = 0; i < intervals.Length; i++)
+                rwl.EnterReadLock();
+                try
+                {
+                    for (var i = 0; i < howMany; i++)
                     {
-                        timingIntervals[i] = new TimingInterval
+                        var epochIndex = epoch % Size;
+                        var interval = intervals[epochIndex];
+
+                        intervalsToFill[i] = new TimingInterval
                         {
-                            IntervalStart = intervals[i]
+                            IntervalStart = GetDateTime(epoch),
                         };
 
-                        MeasurementInterval bucket;
-
-                        if (endpointInstanceData.TryGetValue(intervals[i], out bucket))
+                        // the interval might contain data from the right epoch, or epochs before that have the same index
+                        // we calculate data only if that's the right epoch
+                        if (interval.Epoch == epoch)
                         {
-                            totalDuration += bucket.TotalTime;
-                            totalMeasurements += bucket.TotalMeasurements;
+                            intervalsToFill[i].TotalTime = interval.TotalTime;
+                            intervalsToFill[i].TotalMeasurements = interval.TotalMeasurements;
 
-                            timingIntervals[i].TotalTime = bucket.TotalTime;
-                            timingIntervals[i].TotalMeasurements = bucket.TotalMeasurements;
+                            totalDuration += interval.TotalTime;
+                            totalMeasurements += interval.TotalMeasurements;
                         }
+
+                        epoch -= 1;
```

Comment:
Ah correct. The epoch always counts down, the index gets modded. I had them muddled in my head.